### ISSUE TITLE
replace dc.utils.toHierarchy with d3.stratify

### ISSUE THREAD

**Issue by gordonwoodhull** [dc-js/dc.js#1429]
[As noted by @kum-deepak in #1388](https://github.com/dc-js/dc.js/pull/1388#issuecomment-377445140):

> Can https://github.com/d3/d3-hierarchy/blob/master/README.md#stratify replace dc.utils.toHierarchy?

It is probably more robust. It looks like the main differences are that it doesn't provide the path as one of the members, and it uses `name` instead of `key`. I'm sure `dc.sunburstChart` could adapt to these changes.

The more we use d3 utilities instead of reinventing them, the better off we are, for robustness and interpretability of the code.

**Comment by ialarmedalien** [comment]
I have been looking at this as I've got some other hierarchical graphs that I was thinking of adapting -- the main difference between `dc.utils.toHierarchy` and `stratify` is that the former adds in missing parent nodes, whereas `stratify` throws an error. If `dc.js` is to use `stratify`, there would need to be some little utility that filled in those gaps in the hierarchy, and I'm not sure that there is a way to check for missing parents without constructing the tree!

**Comment by kum-deepak** [comment]
Thanks @ialarmedalien! I can see that root node may be missing in some cases. Are there other cases where parent node might be missing?

**Comment by ialarmedalien** [comment]
Yes -- I think all the examples (in `web/examples`) create "fake" parent nodes.

**Comment by gordonwoodhull** [comment]
Maybe it is more like [d3.nest](https://github.com/d3/d3-collection#nest) with the number of keys dependent on the longest path in the data set. (And with parent nodes annotated with the current path.)

It doesn't look like `dc.utils.toHierarchy` will work correctly if the parents are present, unless the parent is before all of its children. It will blindly append a leaf whenever it runs out of path elements. So maybe it would be good to specify that the parents should *not* be present? I'm not sure.

**Comment by kum-deepak** [comment]
I will suggest few test cases to cover this function. This would allow us to test out any replacement that we may try.

**Comment by ialarmedalien** [comment]
Great -- I have written some code to replace it but haven't yet written any tests.

**Comment by ialarmedalien** [comment]
I have made a set of changes to the sunburst chart in [my fork of dc.js](https://github.com/ialarmedalien/dc.js); I created a [`hierarchyMixin`](https://github.com/ialarmedalien/dc.js/blob/develop/src/hierarchy-mixin.js) which handles converting flat data to a tree structure using `d3.stratify`, and rewrote the existing code to use the tree structure generated by `d3.stratify`. [`chart.formatData()`](https://github.com/ialarmedalien/dc.js/blob/develop/src/hierarchy-mixin.js#L99) takes a flat data array and a d3 layout function; it calls [`chart.stratify()`](https://github.com/ialarmedalien/dc.js/blob/develop/src/hierarchy-mixin.js#L134), which adds in missing parent nodes and runs `d3.stratify` on the data to produce a d3 hierarchy. It then applies the layout function to the tree and returns an array of the nodes in the tree structure. There is also a tiny `partitionMixin`(https://github.com/ialarmedalien/dc.js/blob/develop/src/partition-mixin.js) with partition chart-specific functions as I'm interested in adapting the rectangular partition chart for dc.js.

I also did some work to reduce the large degree of code repetition between the sunburst and pie charts. I have moved the code relating to legend elements into [`legendableMixin`](https://github.com/ialarmedalien/dc.js/blob/develop/src/legendable-mixin.js) and the rest of the shared code into [`pieTypeMixin`](https://github.com/ialarmedalien/dc.js/blob/develop/src/pie-type-mixin.js). Some of the mixin functions had to be overridden in `dc.pieChart` and `dc.sunburstChart`, and there is [a little bit of hackiness](https://github.com/ialarmedalien/dc.js/blob/develop/src/pie-type-mixin.js#L94) to keep some similar-but-not-quite-the-same functions in `pieTypeMixin`. As a result of this merge, the sunburst chart now supports external labels and label paths (like pie charts), and I fixed a bug that meant that the colours in the legend bore no resemblance to those in the sunburst chart. The test suite also passes.

Unfortunately I have introduced a "feature" somewhere that has stopped the sunburst chart from being updated when there are other charts present; clicking on the sunburst will filter it and the other charts, but clicking on the other charts does not filter the sunburst chart. Once I have figured out where the problem is, I will add some cases to the test suite to check for similar issues.
